content/js1/sprints/2/prep/index.md

```diff
@@ -8,4 +8,406 @@ backlog= 'Module-JS1'
+- Explain how to reuse some code to perform a specific task
+- Propose clear and meaningful parameter names for a function
+- Given a problem, suggest clear semantic names for a function designed to solve the given problem
+- Give examples of how functionality can be generalised
+
+## Notes
+
+### 🧩 Percentages
+
+Given a number I want to convert it into a percentage format.
+
+For example, given the number `0.5` we should get `"50%"`
+And given the number `0.23` we should get `"23%"` and so forth.
+
+### Restating the problem
+
+We want to implement some functionality to convert _any_ given number into a percentage.
+Previously, we saw how to use some functions to perform a given task. Here are some examples:
+
+```js {linenos=table,linenostart=1}
+console.log("hello world"); // prints "hello world" to the console
+Math.round(3.141); // will output 3
+```
+
+All these expressions are _function calls_: we’re passing input to the functions, `console.log` or `Math.round` to use their functionality. In the case of `Math.round` and `console.log` these functions have been created and made available to use whenever we start running a program with NodeJS.
+
+However, there is so no such pre-built function that allows us to convert any number to a percentage. So we're going to create a function called `convertToPercentage` with the following properties:
+
+- takes any number input
+- converts it to a string output in percentage form.
+
+```js {linenos=table,linenostart=1}
+convertToPercentage(0.5); // should return "50%"
+```
+
+```js {linenos=table,linenostart=1}
+convertToPercentage(0.23); // should return "23%"
+```
+
+### Useful expressions
+
+It is often helpful to solve a problem in one specific instance before doing it for all cases.
+
+In programming, we always try the simplest thing first. Let’s consider how to convert just one number in to a percentage.
+Let’s consider the variable declaration below:
+
+```js {linenos=table,linenostart=1}
+const num = 0.5;
+```
+
+We want to create an expression for the percentage using the value of `num`.
+To convert to a percentage, we can multiply the number by 100 and then add a `"%"` sign on the end.
+
+So we can write:
+
+```js {linenos=table,linenostart=1}
+const num = 0.5;
+const percentage = `${num * 100}%`;
+```
+
+Recalling template literals, the expression in the curly braces will be evaluated first and then inserted into the string, giving us the percentage string.
+
+{{<note type="exercise" title="Exercise 1.1">}}
+
+The following variables represent the dimensions of a rectangle:
+
+```js
+height; // could have a value like 10
+width; // could have a value like 30
+```
+
+Write expressions for the following in terms of the variables above:
+a) area of the rectangle
+b) perimeter of the rectangle
+
+{{</note>}}
+
+{{<note type="exercise" title="Exercise 1.2">}}
+
+The following variable represents the cost of something in pence:
+
+```js
+price; // could have a value like "130p"
```

Comment:
This feels like a huge jump and requires working out a bunch of string indexing and stuff?

I'm imagining this looking something like (possibly with some number conversions instead):

```js
function toPounds(penceString) {
  // Assume the string ended with a `p`.
  const penceStringWithoutTrailingP = penceString.substring(0, penceString.length - 1);
  const paddedPenceNumberString = penceStringWithoutTrailingP.padStart(3, '0');
  const pounds = paddedPenceNumberString.substring(0, paddedPenceNumberString.length - 2);
  const pence = paddedPenceNumberString.substring(paddedPenceNumberString.length - 2).padEnd(2, '0');
  return `£${pounds}.${pence}`
}
```

Which is a _lot_ of new concepts.

Do you perhaps mean for the values to be numbers instead of strings? Converting `130` to `£1.30` seems a lot more reasonable than converting `"130p"`, with the knowledge from week 1...

Reply:
I think I meant to put 130 here yeah.